### ISSUE TITLE
model.output poor performance

### ISSUE THREAD

**Issue by akhodakivskiy** [deeplearning4j/deeplearning4j#1861]
I'm experiencing poor performance of `model.eval(DataSetIterator)`. This call in my case takes ~5 minutes, while fitting 1 epoch over the same data takes ~40 seconds.

**Comment by AlexDBlack** [comment]
As noted in gitter: we'll need a *lot* more information here.
This is not something I've seen before, ever.
Can you give us some code that reproduces this?

**Comment by akhodakivskiy** [comment]
The code is very simple:

    DataSetIterator it = ...
    mode.eval(it)

Do you want the iterator code, or the model code, or both?

**Comment by AlexDBlack** [comment]
Ideally both. Best option for me is to have something I can copy + paste and click run.

**Comment by akhodakivskiy** [comment]
I'll see if I can isolate this.

**Comment by AlexDBlack** [comment]
@akhodakivskiy also: visualvm or yourkit screenshots could be another option here instead of runnable code

**Comment by akhodakivskiy** [comment]
Sorry, been busy with other stuff. Here are some screenshots

![screenshot from 2016-07-23 00-13-48](https://cloud.githubusercontent.com/assets/145806/17076578/b5ac38d6-506a-11e6-9c96-b007ab934f03.png)

![screenshot from 2016-07-23 00-14-01](https://cloud.githubusercontent.com/assets/145806/17076580/bfab0510-506a-11e6-96bb-62cbc783f49c.png)

![screenshot from 2016-07-23 00-15-40](https://cloud.githubusercontent.com/assets/145806/17076587/c83c8604-506a-11e6-97bf-a9f10f61e800.png)

![screenshot from 2016-07-23 00-14-23](https://cloud.githubusercontent.com/assets/145806/17076584/c53f5eb8-506a-11e6-90c3-dc8a6ad64455.png)

![screenshot from 2016-07-23 00-14-12](https://cloud.githubusercontent.com/assets/145806/17076582/c1e83096-506a-11e6-998e-88ea48021065.png)

**Comment by raver119** [comment]
So that's CUDA backend. Add details about your CUDA environment please. Number of devices, CUDA version (full one), and show the code, how you create **input** data for model.output()

**Comment by AlexDBlack** [comment]
Thanks for the info. Not quite as informative as I would have liked, however.

BTW, are you using time series data (with masking) by any chance?

**Comment by akhodakivskiy** [comment]
It's GTX 980 4 Gb, single unit running 367.27, CUDA 7.5.18, under Ubuntu 16.04.

The DataSetIterator is created like so:

```
RecordReader rr = new CSVRecordReader()
rr.initialize(new CollectionInputSplit(files))
RecordReaderDataSetIterator rrit = new RecordReaderDataSetIterator(rr, new SelfWritableConverter, batchSize, 0, 1, 2, -1, false)

val it = new CachingDataSetIterator(rrit)
```
Then I reuse `it` for normalization, training, and evaluation.

**Comment by akhodakivskiy** [comment]
@AlexDBlack - no time series, no masking. Simple net with 1 DenseLayer, 1 output layer

**Comment by AlexDBlack** [comment]
@akhodakivskiy Thanks, that rules out one possible cause.

@raver119 while I think of it, we probably want to look at improving this:
getRow + argmax in a loop:
https://github.com/deeplearning4j/deeplearning4j/blob/1147f9d1c4034be9b7a9e36e87aae6f91e6097c5/deeplearning4j-core/src/main/java/org/deeplearning4j/eval/Evaluation.java#L187

also this (which is why I was asking about time series + masking):
https://github.com/deeplearning4j/deeplearning4j/blob/1147f9d1c4034be9b7a9e36e87aae6f91e6097c5/deeplearning4j-core/src/main/java/org/deeplearning4j/eval/Evaluation.java#L287

**Comment by agibsonccc** [comment]
Closing this due to inactivity.
